src/lua/uri.c

```diff
@@ -0,0 +1,313 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2021, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#include "lua/uri.h"
+
+#include "lua/utils.h"
+#include "uri/uri.h"
+#include "diag.h"
+
+/**
+ * Add values to URI query parameter @a param from lua table, which located
+ * at the top of the lua stack. All not numeric keys are  treaded as errors.
+ * Allowed types for URI query parameter values are LUA_TSTRING and LUA_TNUMBER.
+ */
+static int
+uri_add_param_from_lua_table(struct uri_query_param *param, struct lua_State *L)
+{
+	int rc = 0;
+	lua_pushnil(L);
+	while (lua_next(L, lua_gettop(L) - 1) != 0 && rc == 0) {
+		if (lua_type(L, -2) == LUA_TNUMBER) {
+			int type = lua_type(L, -1);
+			if (type == LUA_TSTRING || type == LUA_TNUMBER) {
+				const char *val = lua_tostring(L, -1);
```

Comment:
This can be simplified:

```C
int count = lua_rawlen(L, -1);
for (int i = 0; i < count; i++) {
    lua_rawgeti(L, -1, i);
    const char *value = lua_tostring(L, -1)
    if (value != NULL)
        uri_add_query_param(uri, name, value);
    lua_pop(L, 1);
}
```

I think it's okay to silently ignore non-string keys.

Reply:
fixed but i think, that we should check value type, to prevent something like this params = { "x" = function()end}

Reply:
Come to think of it, you're probably right - checking the value type makes sense. Still we can do it with `lua_tostring`:

```C
const char *value = lua_tostring(L, -1)
if (value == NULL)
    /* error */
```

Reply:
as i know lua_tostring same as tostring in lua convert function lua values to string

Reply:
It should be fine. We do it all the time. Or not? I wouldn't bother, but please check other places in our code.

Reply:
Yes you right! I see documentation and you right only for string and number it's return not null